### ISSUE TITLE
Qwen3.5-35B-A3B: KV cache reuse not supported — full prompt recompute on every request

### ISSUE THREAD

**Issue by shashwata2020** [lmstudio-ai/lmstudio-bug-tracker#1563]
**Which version of LM Studio?**
LM Studio 0.4.5 (Build 1)

**Which operating system?**
Windows

**What is the bug?**
When running Qwen3.5-35B-A3B, the llama.cpp backend silently disables KV cache reuse for this model, forcing a full cold prompt recompute on every single request. This results in significantly slower prompt processing (prefill). Multi-turn conversations are especially affected since the shared system prompt and conversation history must be recomputed on every turn.

**Screenshots**
N/A

**Logs**
slot update_slots: id  0 | task 5007 | cache reuse is not supported - ignoring n_cache_reuse = 256
slot update_slots: id  0 | task 5007 | n_tokens = 1560, memory_seq_rm [1560, end)
slot update_slots: id  0 | task 5007 | failed to truncate tokens with position >= 1560 - clearing the memory
slot prompt_clear: id  0 | task 5007 | clearing prompt with 1560 tokens
slot update_slots: id  0 | task 5007 | prompt processing progress, n_tokens = 512, batch.n_tokens = 512, progress = 0.241055

**To Reproduce**
1. Load Qwen3.5-35B-A3B (GGUF) in LM Studio 0.4.5
2. Send any prompt
3. Open Developer Logs
4. Observe "cache reuse is not supported" in the logs

**Comment by Jockel12** [comment]
Same Issue here with the 122b model in Q3

**Comment by SquishedSquirrel** [comment]
I'm getting full prompt recompute on any model that approaches 100% KV size. No idea if they are related. Went back to 0.3.39 until release notes mention anything that might be a fix.

**Comment by shashwata2020** [comment]
Aparently llama.cpp does not support qwen 3/3.5 multimodal model kv cache reuse yet. But its a work in progress, hopefully soon!

https://github.com/ggml-org/llama.cpp/pull/19849
https://github.com/ggml-org/llama.cpp/pull/19877

**Comment by bunnynode** [comment]
mlx 0.30.6 prompt cache still not work with qwen35

**Comment by siriux** [comment]
mlx related issue here: https://github.com/ml-explore/mlx-lm/issues/903

**Comment by LisaMacintosh** [comment]
using runtime v2.5.0 and still not fixed unfortunately

**Comment by iseeboys** [comment]
still not fixed.. the fix is there, but LM Studio hasn't put it in yet??

**Comment by drHuangMHT** [comment]
I can confirm that in llama.cpp `b8212` the caching is now working with significant TTFT reduction on consecutive conversations, model `Qwen3.5-122B-A10B`. Previously broken in `b8131` with recomputing of KV cache on every round.

**Comment by iseeboys** [comment]
> I can confirm that in llama.cpp `b8212` the caching is now working with significant TTFT reduction on consecutive conversations, model `Qwen3.5-122B-A10B`. Previously broken in `b8131` with recomputing of KV cache on every round.

thank you for the heads up! seriously! <3

**Comment by celestial-rose** [comment]
is this supposed to be fixed ? 
I still get `cache reuse is not supported - ignoring n_cache_reuse = 256` using Qwopus V3 27B Q8 GGUF; current `llama.cpp release b8656` 

EDIT: unsure about the `cache reuse is not supported` since I actually get 
```
[2026-04-05 23:58:25][DEBUG] LlamaV4::predict slot selection: session_id=<empty> server-selected (LCP/LRU)
[2026-04-05 23:58:26][DEBUG] slot get_availabl: id  3 | task -1 | selected slot by LCP similarity, sim_best = 0.999 (> 0.100 thold), f_keep = 0.999
[2026-04-05 23:58:26][DEBUG] slot launch_slot_: id  3 | task -1 | sampler chain: logits -> ?penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> top-p -> ?min-p -> ?xtc -> temp-ext -> dist 
slot launch_slot_: id  3 | task 37927 | processing task, is_child = 0
slot update_slots: id  3 | task 37927 | new prompt, n_ctx_slot = 262144, n_keep = 10649, task.n_tokens = 121327
slot update_slots: id  3 | task 37927 | cache reuse is not supported - ignoring n_cache_reuse = 256
slot update_slots: id  3 | task 37927 | n_past = 121176, slot.prompt.tokens.size() = 121306, seq_id = 3, pos_min = 121305, n_swa = 0
slot update_slots: id  3 | task 37927 | Checking checkpoint with [121173, 121173] against 121176...
[2026-04-05 23:58:26][DEBUG] slot update_slots: id  3 | task 37927 | restored context checkpoint (pos_min = 121173, pos_max = 121173, n_tokens = 121174, n_past = 121174, size = 149.626 MiB)
slot update_slots: id  3 | task 37927 | n_tokens = 121174, memory_seq_rm [121174, end)
[2026-04-05 23:58:26][DEBUG] slot update_slots: id  3 | task 37927 | prompt processing progress, n_tokens = 121323, batch.n_tokens = 149, progress = 0.999967
[2026-04-05 23:58:26][DEBUG] slot update_slots: id  3 | task 37927 | n_tokens = 121323, memory_seq_rm [121323, end)
[2026-04-05 23:58:26][INFO][qwopus3.5-27b-v3@q8_0] Prompt processing progress: 0.0%
```

**Comment by tarunkapadia93** [comment]
i am still getting it with Qwen 3.5 9B 27B and the bigger MoE model too 

Why is the log from 9B 
build_info: b8744-d7ff074c8

```
slot launch_slot_: id  3 | task 2739 | processing task, is_child = 0
slot update_slots: id  3 | task 2739 | new prompt, n_ctx_slot = 131072, n_keep = 0, task.n_tokens = 74667
**slot update_slots: id  3 | task 2739 | cache reuse is not supported - ignoring n_cache_reuse = 512**
slot update_slots: id  3 | task 2739 | n_past = 74108, slot.prompt.tokens.size() = 74247, seq_id = 3, pos_min = 74246, n_swa = 0
slot update_slots: id  3 | task 2739 | Checking checkpoint with [74035, 74035] against 74108...
slot update_slots: id  3 | task 2739 | restored context checkpoint (pos_min = 74035, pos_max = 74035, n_tokens = 74036, n_past = 74036, size = 50.251 MiB)
slot update_slots: id  3 | task 2739 | n_tokens = 74036, memory_seq_rm [74036, end)
slot update_slots: id  3 | task 2739 | prompt processing progress, n_tokens = 74151, batch.n_tokens = 115, progress = 0.993089
slot update_slots: id  3 | task 2739 | n_tokens = 74151, memory_seq_rm [74151, end)
slot update_slots: id  3 | task 2739 | prompt processing progress, n_tokens = 74663, batch.n_tokens = 512, progress = 0.999946
slot update_slots: id  3 | task 2739 | created context checkpoint 27 of 128 (pos_min = 74150, pos_max = 74150, n_tokens = 74151, size = 50.251 MiB)
slot update_slots: id  3 | task 2739 | n_tokens = 74663, memory_seq_rm [74663, end)
reasoning-budget: activated, budget=2147483647 tokens
reasoning-budget: deactivated (natural end)
```

**Comment by fevaoctwh61-web** [comment]
LM Studio 0.4.10+1
Metal llama.cpp v2.13.0
Qwen3.5 35B A3B

2026-04-16 19:55:56 [DEBUG]
 slot launch_slot_: id  3 | task -1 | sampler chain: logits -> penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> top-p -> ?min-p -> ?xtc -> temp-ext -> dist 
slot launch_slot_: id  3 | task 6456 | processing task, is_child = 0
slot update_slots: id  3 | task 6456 | new prompt, n_ctx_slot = 262144, n_keep = 0, task.n_tokens = 130190
2026-04-16 19:55:56 [DEBUG]
 slot update_slots: id  3 | task 6456 | cache reuse is not supported - ignoring n_cache_reuse = 256
slot update_slots: id  3 | task 6456 | n_tokens = 47716, memory_seq_rm [47716, end)

**Comment by MiheevN** [comment]
I don't know if this is a bug, but the same thing happens on qwen3.6-35b-a3b and LM Studio
0.4.12 (Build 1)

**Comment by sanderramaekers-dotcom** [comment]
Same here on LM Studio (Version 0.4.12+1), qwen3.5-9b

**Comment by johannesrave** [comment]
also from LM Studio Version 0.4.12+1

2026-05-08 08:50:12  [INFO]
 [qwen/qwen3.6-35b-a3b] Running chat completion on conversation with 75 messages.
2026-05-08 08:50:12  [INFO]
 [qwen/qwen3.6-35b-a3b] Streaming response...
2026-05-08 08:50:12 [DEBUG]
 LlamaV4::predict slot selection: session_id=<empty> server-selected (LCP/LRU)
2026-05-08 08:50:12 [DEBUG]
 slot get_availabl: id  0 | task -1 | selected slot by LCP similarity, sim_best = 0.995 (> 0.100 thold), f_keep = 1.000
2026-05-08 08:50:12 [DEBUG]
 slot launch_slot_: id  0 | task -1 | sampler chain: logits -> ?penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> ?top-p -> ?min-p -> ?xtc -> temp-ext -> dist 
slot launch_slot_: id  0 | task 113340 | processing task, is_child = 0
slot update_slots: id  0 | task 113340 | new prompt, n_ctx_slot = 128000, n_keep = 30995, task.n_tokens = 95809
slot update_slots: id  0 | task 113340 | cache reuse is not supported - ignoring n_cache_reuse = 256
slot update_slots: id  0 | task 113340 | n_tokens = 95342, memory_seq_rm [95342, end)
slot update_slots: id  0 | task 113340 | prompt processing progress, n_tokens = 95805, batch.n_tokens = 463, progress = 0.999958
2026-05-08 08:50:12 [DEBUG]
 slot update_slots: id  0 | task 113340 | created context checkpoint 16 of 32 (pos_min = 95341, pos_max = 95341, n_tokens = 95342, size = 62.813 MiB)
2026-05-08 08:50:12  [INFO]
 [qwen/qwen3.6-35b-a3b] Prompt processing progress: 0.0%
2026-05-08 08:50:12 [DEBUG]
 slot update_slots: id  0 | task 113340 | n_tokens = 95805, memory_seq_rm [95805, end)
2026-05-08 08:50:12 [DEBUG]
 slot init_sampler: id  0 | task 113340 | init sampler, took 7.38 ms, tokens: text = 95809, total = 95809
slot update_slots: id  0 | task 113340 | prompt processing done, n_tokens = 95809, batch.n_tokens = 4
2026-05-08 08:50:12  [INFO]
 [qwen/qwen3.6-35b-a3b] Prompt processing progress: 99.1%
2026-05-08 08:50:14 [DEBUG]
 slot update_slots: id  0 | task 113340 | created context checkpoint 17 of 32 (pos_min = 95804, pos_max = 95804, n_tokens = 95805, size = 62.813 MiB)
2026-05-08 08:50:14  [INFO]
 [qwen/qwen3.6-35b-a3b] Prompt processing progress: 100.0%
2026-05-08 08:50:42 [DEBUG]
 slot print_timing: id  0 | task 113340 | 
prompt eval time =    2324.21 ms /   467 tokens (    4.98 ms per token,   200.93 tokens per second)
       eval time =   27482.15 ms /   959 tokens (   28.66 ms per token,    34.90 tokens per second)
      total time =   29806.36 ms /  1426 tokens
2026-05-08 08:50:42 [DEBUG]
 slot      release: id  0 | task 113340 | stop processing: n_tokens = 96767, truncated = 0
srv  update_slots: all slots are idle
2026-05-08 08:50:42 [DEBUG]
 LlamaV4: server assigned slot 0 to task 113340

**Comment by takeseem** [comment]
[Qwen3.6-35B-A3B-Q4_K_M.gguf](https://lmstudio.ai/models/qwen/qwen3.6-35b-a3b)

It's same issues. but it's work in llama-server vulkan.
```
llama-server --version
version: 9334 (192d8ae8b)
```

2026-05-27 update llama.cpp vulkan to `- llama.cpp release b9334 (commit 5190c2e)` 

it's work. thanks for LM Studio update.

**Comment by muschaw** [comment]
Same issue testing with Qwen3.6-27B (MTP) MTP is working well, but the cache is not being reused.

`cache reuse is not supported - ignoring n_cache_reuse`

Using the latest b9334 (v2.16.0) on LM Studio 0.4.14 (Build 4)

**Comment by ghost** [comment]
Adding another data point: this also happens with a newer MTP model, `qwen3.6-35b-a3b-mtp`.

When running a chat completion, LM Studio appears to select a slot and update the prompt cache, but then disables cache reuse for the request:

```text
[qwen3.6-35b-a3b-mtp:2] Running chat completion on conversation with 2 messages.
2026-06-14 13:31:16 [INFO]
[qwen3.6-35b-a3b-mtp:2] Streaming response...

2026-06-14 13:31:16 [DEBUG]
LlamaV4::predict slot selection: session_id=<empty> server-selected (LCP/LRU)

0.18.184.407 I slot get_availabl: id  3 | task -1 | selected slot by LRU, t_last = -1
0.18.184.412 I srv  get_availabl: updating prompt cache
0.18.184.416 I srv          load:  - looking for better prompt, base f_keep = -1.000, sim = 0.000
0.18.184.419 I srv        update:  - cache state: 0 prompts, 0.000 MiB (limits: 8192.000 MiB, 64000 tokens, 8589934592 est)
0.18.184.421 I srv  get_availabl: prompt cache update took 0.01 ms

0.18.184.520 I slot launch_slot_: id  3 | task 0 | processing task, is_child = 0
0.18.184.533 W slot update_slots: id  3 | task 0 | cache reuse is not supported - ignoring n_cache_reuse = 256
```

So this looks like it may not be limited to `Qwen3.5-35B-A3B`; it may affect Qwen MTP / A3B-style models more generally.

Expected behavior: LM Studio should support prompt/KV cache reuse for MTP models when possible, or clearly document/show in the UI that cache reuse is currently disabled for MTP models.

Impact: multi-turn chats and repeated prompts are slower because the shared prefix/context cannot be reused and has to be recomputed.

**Comment by titansonico10** [comment]
**Also reproducing with Qwen3-30B-A3B-Instruct in LM Studio**

**Setup:**
- LM Studio (latest)
- Windows
- GPU: 16 GB VRAM
- RAM: 64 GB
- Model: `Qwen3-30B-A3B-Instruct` (GGUF)
- Used via Cline (VS Code extension) with a long system prompt (~400-500 tokens)

**Observed behavior:**
Every single turn, the logs show `cache reuse is not supported - ignoring n_cache_reuse`, which means the full system prompt and conversation history are recomputed from scratch on each request. This makes multi-turn conversations via Cline noticeably slow on the prefill phase, even though the conversation prefix hasn't changed.

This seems to be a general issue affecting all MoE/sparse A3B-style models where llama.cpp silently disables KV cache reuse, not just Qwen3.5-35B-A3B.

**Comment by shashwata2020** [comment]
Qwen models with image modality does not use cache reuse it uses checkpoints, so im closing this issue.

**Comment by ghost** [comment]
I think this issue may have been closed based on a misunderstanding.

Several people in this thread are not using Qwen models with image modality. The reports include text-only / non-vision models such as:

- `Qwen3.5-35B-A3B`
- `Qwen3.5-122B-A10B`
- `qwen3.5-9b`
- `Qwen3.6-35B-A3B`
- `Qwen3-30B-A3B-Instruct`
- `qwen3.6-35b-a3b-mtp`

So the explanation "Qwen models with image modality do not use cache reuse" does not seem to address the actual reports here.

The common symptom people are reporting is this log line on text/A3B/MoE-style models:

```text
cache reuse is not supported - ignoring n_cache_reuse
```

and the impact is slower multi-turn usage because the shared prompt / conversation prefix is not being reused the way users expect.

If LM Studio is intentionally using checkpoints instead of `n_cache_reuse` for these models, could you please clarify that explicitly and document whether this log line is expected/harmless? If not, I think this issue should be reopened, because multiple users are reproducing it on non-image models and the closing comment does not match the examples posted above.
